source/extensions/filters/network/thrift_proxy/protocol.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/filters/network/thrift_proxy/protocol.h"
+
+#include <algorithm>
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+#include "common/common/byte_order.h"
+#include "common/common/macros.h"
+
+#include "extensions/filters/network/thrift_proxy/binary_protocol.h"
+#include "extensions/filters/network/thrift_proxy/buffer_helper.h"
+#include "extensions/filters/network/thrift_proxy/compact_protocol.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+bool AutoProtocolImpl::readMessageBegin(Buffer::Instance& buffer, std::string& name,
+                                        MessageType& msg_type, int32_t& seq_id) {
+  if (protocol_ == nullptr) {
+    if (buffer.length() < 2) {
+      return false;
+    }
+
+    uint16_t version = BufferHelper::peekU16(buffer);
+    if (BinaryProtocolImpl::isMagic(version)) {
+      setProtocol(std::make_unique<BinaryProtocolImpl>());
```

Comment:
Is it possible to determine at this point in the processing whether we need a `LaxBinaryProtocolImpl`?

Reply:
The only guarantee that the non-strict binary protocol makes is that the leading bit is 0, so misidentifying another protocol seems pretty likely and the result would be interpreting the first four bytes as message name length. For instance, with auto-transport and auto-protocol, an errant HTTP request would be interpreted as an unframed, non-strict binary thrift message with a message name of ~1.1 GB.

I think longer term, we'll want to have a configurable limit on the size of encoded strings (perhaps separate limits for message names vs. fields). With that in place, I'd feel more comfortable allowing the non-strict binary protocol to be detected automatically.